Pipeline creator mk2

HarryDC opened the pull request:
This updates the SlicerPipelines module to utilize the ParameterWrapper infrastructure. 

While most of the changes are internal there are some functional changes as well

In Pipeline Creator, the user doesn't chose the type for an input parameter anymore, it is determined when the input is connected to one of the pipeline parameters. This allows us to carry annotations from the pipeline steps forward into the newly created pipeline.

Pipeline Case Iterator doesn't read files from a directory anymore but is now driven via a `.csv` file that determines the input parameters for each "pass", it can also handle multiple output parameters and will store the results as files (for nodes) as well as in a output `.csv` file 

Still needs more documentations with regards to Use of Creator, and Case iterator and Pipeline Construction also see https://github.com/Slicer/Slicer/issues/7015

HarryDC commented:
Added documentation in writing i found a couple of issues i still need to check/resolve

- Iterator 
  - [ ] Iterator does not clear old .csv file when changing the pipeline
  - [ ] User can't configure output file name
  - [ ] Check that relative paths work for input in .csv file
  - [ ] I think "Output Extension" has become superflous, need to check

HarryDC commented:
@jcfr @sjh26 Quick User Test ...

Both the pipeline creator and case iterator feature a quickstart in the "Help" menu, it's probably easiest to follow that but here is a test sequence ... . 
 
## Pipeline Creator 

1. Open pipeline creator
2. Use "ScaleAndTransform" for the pipeline name and a directory of your chosing as the output dir
3. In "Step 0 - Inputs"
    1. Add 2 more input parameters by clicking '+' in the "Step 0 - Inputs" 
    2. Label the parameters `mesh`, `scalefactor`, `translateFactor` respectively
4. Click on "Add Step", Select `SurfaceToolBox.ScaleMesh`
5. Under `Step 1 - SurfaceToolBox.ScaleMesh` 
   1. In the combobox next to mesh select `step0_None_mesh` 
   2. Click on the "Fixed" checkbox next to scaleX and scaleY and select `Step0_None_scalefactor` in the combobox
   3. In the box scaleZ enter 2.0 or another value you like
6. Click on "Add Step", Select `SurfaceToolbox.TranslateMesh`
7. Under `Step 2 - SurfaceToolBox.TranslateMesh`
    1. In the combobox  next to mesh select `step1_SurfaceToolbox.ScaleMesh_return` (this connects the output of the previous step to the input of this step
    2. Clicked on the "Fixed" checkbox next to translateX and translateY and select `Step0_None_translatefactor` in the combobox
    3. In the box translateZ enter 2.0 or any other value you like
8. In "Step 3 Outputs"
    1. Click on "+" to create another output
    2. Name the first output  "scaled" and select `step1_SurfaceToolbox.ScaleMesh_return` in the combobox
    3. Name the second output "scaletranslated" and select `step2_SurfaceToolbox.TranslateMesh_return`
 9. In the top section click on "Test" This should show the pipeline interface and creates a pipeline that will first scale and then translate the input model.
 10. Click on "Back" and "Generate" to write out the completed module source, if you left "Load Module on creation" on you should have a completed pipeline under `PipelineModules` in the modules dropdown, open up and check the interface it should look somthing like this 
![image](https://github.com/KitwareMedical/SlicerPipelines/assets/1878111/63956a09-3e50-49fa-a40a-599e0eb6ef23)

 
## Case Iterator

Select the case iterator module
1. Next to pipeline push "Select" and select the pipeline from above `ScaleAndTranslate`
2. Click on "Create Template File" and save a file in an appropriate location, line first line will looks like 
```
model:vtkMRMLModelNode,scaleFactor:float,translateFactor:float
```
3. Edit the file with paths to models and factors that you like, paths may be relative to the location of the csv file 
4. Select a directory for the outputs
5. Click on "Run" After the pipeline executes there should be 2 output files for each input file, one just scaled, the other one translated and scaled with the params that you provided 


I hope this helps evaluating this, thanks !

allemangD commented on PipelineCaseIterator/PipelineCaseIterator.py:
```diff
@@ -4,300 +4,493 @@
-from PipelineCreatorLib.Asynchrony import Asynchrony
-from PipelineModulesLib.Util import ScopedNode, ScopedDefaultStorageNode, human_sorted
+from PipelineCreator import PipelineCreatorLogic, PipelineProgressCallback
+from PipelineCaseIteratorLibrary import *
```
This fails on cmake install since `PipelineCaseIteratorLibrary/__init__.py` isn't included in `PipelineCaseIterator/CMakeLists.txt`, so `import *` grabs the child modules instead.

Leads to strange NameErrors later on in the code, `Asynchrony` and `IteratorParameterFile` are missing.

I suggest adding an `__all__` to `PipelineCaseIteratorLibrary/__init__.py` and replacing these `import *` usages with explicit lists as in lines 10-14 to more easily catch this kind of error in the future.

allemangD replied:
I pushed 6d44c90 for those first two points, but I left `import *` in since that's more invasive and I'm not entirely sure where all the usages of those classes are.

For reference, the steps to reproduce the issue:

```
gh repo clone KitwareMedical/SlicerPipelines
cd SlicerPipelines
cmake -B build
cmake --build build
build/SlicerWithPipelines
```

Then follow the guide you listed in https://github.com/KitwareMedical/SlicerPipelines/pull/23#issuecomment-1679488895

allemangD commented on PipelineCaseIterator/Testing/Python/CMakeLists.txt:
The default-generated qSlicerPipelineCaseIteratorModuleGenericTest fails when saving settings, I suppose because the generic test runs headless and the UI is not created?

```
Traceback (most recent call last):
  File "./bin/Python/slicer/ScriptedLoadableModule.py", line 133, in _onModuleAboutToBeUnloaded
    self.cleanup()
  File "./lib/Slicer-5.4/qt-scripted-modules/PipelineCaseIterator.py", line 481, in cleanup
    settings.setValue('PipelineCaseIterator/LastResultsFileName', self.ui.resultsFileNameLabel.text)
AttributeError: '' object has no attribute 'resultsFileNameLabel'
```

allemangD commented on PipelineCaseIterator/PipelineCaseIterator.py:
```diff
@@ -4,300 +4,493 @@
+    valid = True
+
+    if baseDirectory == '':
+        baseDirectory = os.curdir()
```
This should be `os.getcwd()`. Causes CaseIteratorRunnerTest to fail.

allemangD commented on PipelineCaseIterator/Testing/Python/CaseIteratorRunnerTest.py:
```diff
@@ -0,0 +1,53 @@
+    def testSimpleRowToTypes(self):
+        row = {"a": "1", "b": "2", "node": self.testFileName }
+        types = {"a": str, "b": int, "node": slicer.vtkMRMLScalarVolumeNode }
+        data, nodes = rowToTypes(row, types)
```
too many values to unpack

allemangD commented on PipelineCaseIterator/Testing/Python/CaseIteratorRunnerTest.py:
```diff
@@ -0,0 +1,53 @@
+    def testConversionError(self):
+        row = {"a": "1", "b": "this is no int", "node": self.testFileName }
+        types = {"a": str, "b": int, "node": slicer.vtkMRMLScalarVolumeNode }
+        data, nodes = rowToTypes(row, types)
```
too many values to unpack

allemangD commented on PipelineCaseIterator/Testing/Python/CaseIteratorRunnerTest.py:
```diff
@@ -0,0 +1,53 @@
+    def testLoadingError(self):
+        row = {"a": "1", "b": "2", "node": "filedoesnotexist.nrrd" }
+        types = {"a": str, "b": int, "node": slicer.vtkMRMLScalarVolumeNode }
+        data, nodes = rowToTypes(row, types)
```
too many values to unpack

allemangD commented on PipelineCaseIterator/PipelineCaseIterator.py:
```diff
@@ -4,300 +4,493 @@
+    Args:
+        csvRow (dict[str, str]): The row from the csv file
+        inputTypes (dict[str, typing.Any]): The types of the input parameters from the pipelineInfo
+    Returns:
```
Now out-of-date since it returns `valid` also.

allemangD reviewed (changes requested):
After 6d44c90a32de9bf2b794899f3ce6cc53dbc8bae0 I'm able to build, run through the guide, and test without issues. I haven't review _all_ the code, there's a lot there, but what I have looked through all makes sense.

Some tests do fail; I've added comments on those files.

HarryDC commented:
@allemangD I think i fixed the issues that you reported, @vicory  reported another issue when testing i fixed that as well, it would be great if you could have another look. Thanks !

allemangD reviewed (approved):
Thanks for addressing those issues @HarryDC. Looks ready to merge.
